internal/driver/luks.go

```diff
@@ -115,179 +169,70 @@ func getLuksContext(secrets map[string]string, context map[string]string, lifecy
-		"-v",
-		"--batch-mode",
-		"--cipher", ctx.EncryptionCipher,
-		"--key-size", ctx.EncryptionKeySize,
-		"--key-file", "-",
-		"luksFormat", source,
-	}
-
-	klog.V(4).Info("executing cryptsetup luksFormat command ", cryptsetupArgs)
-
-	cmd := e.Exec.Command(cryptsetupCmd, cryptsetupArgs...)
-
-	// Set the encryption key as stdin
-	cmd.SetStdin(strings.NewReader(ctx.EncryptionKey))
-
-	out, err := cmd.CombinedOutput()
-	if err != nil {
-		return fmt.Errorf("cryptsetup luksFormat failed: %w cmd: '%s %s' output: %q",
-			err, cryptsetupCmd, strings.Join(cryptsetupArgs, " "), string(out))
-	}
-
-	// open the luks partition
-	klog.V(4).Info("luksOpen ", source)
-	err = e.luksOpen(ctx, source)
+func (e *Encryption) luksFormat(ctx LuksContext, source string) (string, error) {
+	luks2 := cryptsetup.LUKS2{SectorSize: 512}
+	keySize, err := strconv.Atoi(ctx.EncryptionKeySize)
 	if err != nil {
-		return fmt.Errorf("cryptsetup luksOpen failed: %w cmd: '%s %s' output: %q",
-			err, cryptsetupCmd, strings.Join(cryptsetupArgs, " "), string(out))
-	}
-
-	defer func() {
-		if e := e.luksClose(ctx.VolumeName); e != nil {
-			klog.Errorf("cannot close luks device: %s", e.Error())
-		}
-	}()
-
-	klog.V(4).Info("The LUKS volume name is ", ctx.VolumeName)
-
-	return nil
-}
-
-// prepares a luks-encrypted volume for mounting and returns the path of the mapped volume
-func (e *Encryption) luksPrepareMount(ctx LuksContext, source string) (string, error) {
-	if err := e.luksOpen(ctx, source); err != nil {
-		return "", err
+		return "", fmt.Errorf("keysize str to int coversion: %w", err)
 	}
-	return "/dev/mapper/" + ctx.VolumeName, nil
-}
-
-func (e *Encryption) luksClose(volume string) error {
-	cryptsetupCmd, err := e.getCryptsetupCmd()
-	if err != nil {
-		return err
+	cipherString := strings.SplitN(ctx.EncryptionCipher, "-", 2)
+	genericParams := cryptsetup.GenericParams{
+		Cipher:        cipherString[0],
+		CipherMode:    cipherString[1],
+		VolumeKeySize: keySize / 8,
 	}
-	cryptsetupArgs := []string{"--batch-mode", "close", volume}
-
-	klog.V(4).Info("executing cryptsetup close command")
+	klog.V(4).Info("Initalizing device to perform luks format ", source)
 
-	out, err := e.Exec.Command(cryptsetupCmd, cryptsetupArgs...).CombinedOutput()
+	newLuksDevice, err := NewLuksDevice(e.CryptSetup, source)
 	if err != nil {
-		return fmt.Errorf("removing luks mapping failed: %w cmd: '%s %s' output: %q",
-			err, cryptsetupCmd, strings.Join(cryptsetupArgs, " "), string(out))
+		return "", fmt.Errorf("initializing luks device to format: %w", err)
 	}
-	return nil
-}
-
-func (e *Encryption) luksOpen(ctx LuksContext, volume string) error {
-	// check if the luks volume is already open
-	if _, err := e.FileSystem.Stat("/dev/mapper/" + ctx.VolumeName); !e.FileSystem.IsNotExist(err) {
-		klog.V(4).Infof("luks volume is already open %s", volume)
-		return nil
-	}
-
-	cryptsetupCmd, err := e.getCryptsetupCmd()
+	klog.V(4).Info("Formatting luks device ", newLuksDevice.device)
```

Comment:
Could you replace the klog with new logger implementation in `pkg/logger`

Reply:
The new implementation requires the context if I am not mistaken. Do you think we should add the context as an argument to `luksFormat` and `luksClose`?

Reply:
Yup, we should add the context and get the logger from it. This way we are able to sort of stack trace in the logs whenever a grpc func is called!